src/main/sensors/compass.c

```diff
@@ -54,35 +60,112 @@ mag_t mag;                   // mag access functions
+// Strategy is
+// 1. If SPI device is defined, it will take precedence, assuming it's onboard.
+// 2. I2C devices are will be handled by address = 0 (per device default).
+// 3. Slave I2C device on SPI gyro
+
+#if defined(USE_SPI) && (defined(USE_MAG_SPI_HMC5883) || defined(USE_MAG_SPI_AK8963))
+    compassConfig->mag_bustype = BUSTYPE_SPI;
+#ifdef USE_MAG_SPI_HMC5883
+    compassConfig->mag_spi_device = SPI_DEV_TO_CFG(spiDeviceByInstance(HMC5883_SPI_INSTANCE));
+    compassConfig->mag_spi_csn = IO_TAG(HMC5883_CS_PIN);
+#else
+    compassConfig->mag_spi_device = SPI_DEV_TO_CFG(spiDeviceByInstance(AK8963_SPI_INSTANCE));
+    compassConfig->mag_spi_csn = IO_TAG(AK8963_CS_PIN);
+#endif
+    compassConfig->mag_i2c_device = I2C_DEV_TO_CFG(I2CINVALID);
+    compassConfig->mag_i2c_address = 0;
+#elif defined(USE_MAG_HMC5883) || defined(USE_MAG_AK8975) || (defined(USE_MAG_AK8963) && !(defined(USE_GYRO_MPU6050) || defined(USE_GYRO_SPI_MPU9250)))
+    compassConfig->mag_bustype = BUSTYPE_I2C;
+    compassConfig->mag_i2c_device = I2C_DEV_TO_CFG(MAG_I2C_INSTANCE);
+    compassConfig->mag_i2c_address = 0;
+    compassConfig->mag_spi_device = SPI_DEV_TO_CFG(SPIINVALID);
+    compassConfig->mag_spi_csn = IO_TAG_NONE;
+#elif defined(USE_MAG_AK8963) && ((defined(USE_GYRO_MPU6050) || defined(USE_GYRO_SPI_MPU9250)))
+    compassConfig->mag_bustype = BUSTYPE_SLAVE;
+    compassConfig->mag_i2c_device = I2C_DEV_TO_CFG(I2CINVALID);
+    compassConfig->mag_i2c_address = 0;
+    compassConfig->mag_spi_device = SPI_DEV_TO_CFG(SPIINVALID);
+    compassConfig->mag_spi_csn = IO_TAG_NONE;
+#else
+    compassConfig->mag_hardware = MAG_NONE;
+    compassConfig->mag_bustype = BUSTYPE_NONE;
+    compassConfig->mag_i2c_device = I2C_DEV_TO_CFG(I2CINVALID);
+    compassConfig->mag_i2c_address = 0;
+    compassConfig->mag_spi_device = SPI_DEV_TO_CFG(SPIINVALID);
+    compassConfig->mag_spi_csn = IO_TAG_NONE;
+#endif
+    compassConfig->interruptTag = COMPASS_INTERRUPT_TAG;
+}
 
-#ifdef MAG
+#if defined(MAG)
 
 static int16_t magADCRaw[XYZ_AXIS_COUNT];
 static uint8_t magInit = 0;
 
-bool compassDetect(magDev_t *dev, magSensor_e magHardwareToUse)
+#if !defined(SITL)
+
+bool compassDetect(magDev_t *dev)
 {
     magSensor_e magHardware;
 
-retry:
+    busDevice_t *busdev = &dev->busdev;
+
+    switch (compassConfig()->mag_bustype) {
+#ifdef USE_I2C
+    case BUSTYPE_I2C:
+        busdev->bustype = BUSTYPE_I2C;
+        busdev->busdev_u.i2c.device = I2C_CFG_TO_DEV(compassConfig()->mag_i2c_device);
+        busdev->busdev_u.i2c.address = compassConfig()->mag_i2c_address;
+#endif
+        break;
+
+#ifdef USE_SPI
+    case BUSTYPE_SPI:
+        busdev->bustype = BUSTYPE_SPI;
+        spiBusSetInstance(busdev, spiInstanceByDevice(SPI_CFG_TO_DEV(compassConfig()->mag_spi_device)));
+        busdev->busdev_u.spi.csnPin = IOGetByTag(compassConfig()->mag_spi_csn);
+#endif
+        break;
+
+#if defined(USE_MAG_AK8963) && ((defined(USE_GYRO_MPU6050) || defined(USE_GYRO_SPI_MPU9250)))
```

Comment:
Likely USE_GYRO_MPU6050 should be USE_GYRO_SPI_MPU6500?

Reply:
With this change the driver is working fro the MPU9250 at the AFNG F7 development environment.